### PR TITLE
Walk/Visitor should follow Enter/Exit Pattern

### LINKED ISSUE #257: Detect leaving Node during ast.Walk
Currently, it appears `ast.Walk` is supposed to work as follows:

- It encounters a node, and it calls `Visit`, passing in the node interface value
- Once it has processed the node, it again calls `Visit` with a `nil` node interface value
  - ... indirectly via a recursive call to `Walk(v, nil)` at the end of `Walk()`

Here's the godocs for the `Visitor` interface, which also describes this intended design:

```go
// Visitor Visit method is invoked for each node encountered by Walk.
// If the result visitor w is not nil, Walk visits each of the children
// of node with the visitor v, followed by a call of v.Visit(nil).
```

First of all, the first few lines of `ast.Walk` prevent it from behaving that way:

```go
	if n == nil {
		return
	}
	if v = v.Visit(n); v == nil {
		return
	}
```

If it sees a nil interface value, it doesn't call `v.Visit()`. I think the intended code there was supposed to be:

```
	if v = v.Visit(n); v == nil {
		return
	}
	if n == nil {
		return
	}
```

Things now seem better. So I wrote a simple program with a Visit that:

- When node is not nil, push the node into a stack
- When node is nil, pop the last node out of the stack
- Panic if we try to pop from an empty stack

Here's the program:

```go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"

	"github.com/robertkrimen/otto/ast"
	"github.com/robertkrimen/otto/parser"
)

var (
	enter int
	leave int
)

type walker []ast.Node

func (w *walker) push(n ast.Node) {
	(*w) = append((*w), n)

	fmt.Println("ENTER:", n)
	enter++
}

func (w *walker) pop() {
	size := len(*w)
	if size <= 0 {
		panic("pop of empty stack")
	}

	fmt.Println("LEAVE:", (*w)[size-1])
	leave++

	(*w)[size-1] = nil
	*w = (*w)[:size-1]
}

func (w *walker) Visit(n ast.Node) ast.Visitor {
	if n == nil {
		w.pop()
	} else {
		w.push(n)
	}

	return w
}

func main() {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("PANIC:", r)
			fmt.Println(enter, leave)
		}
	}()

	src, err := ioutil.ReadFile(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}

	program, err := parser.ParseFile(nil, os.Args[1], string(src), parser.IgnoreRegExpErrors)
	if err != nil {
		log.Fatal(err)
	}

	ast.Walk(new(walker), program)

	fmt.Println(enter, leave)
}
```

... and I've found that I can make this program panic as a result of a pop from an empty stack! It happens when running this code on a [minified JS file](https://github.com/ereyes01/antlr-go-mem/blob/master/bad.js) though I'm sure there's plenty of other ways we can hit the problem.

This is probably happening as follows: in going through the various AST parts, it encounters a nil interface placed there by the parser perhaps, and then passes it into `Visit()` with the intention of "pushing" the Node onto the stack.

Since apparently a nil interface can be encountered in a normal AST, perhaps using nil to denote the "exiting" of the node's subtree in the AST isn't a good idea.

The way I would propose resolving this is to take an explicit enter/exit approach, as is implemented in this new `Visitor` interface:

```go
type Visitor interface {
    Enter(n Node) Visitor
    Exit(n Node)
}
```

... and at the top of `ast.Walk`, we might do something like:

```go
	if n == nil {  // safe to ignore nil Nodes?
		return
	}
	if v = v.Enter(n); v == nil {
		return
	}

       defer v.Exit(n)
```

... And we would remove the recursive `Walk(v, nil)` at the end of `Walk()`.

The downside to this approach is that you would lose the nice `WalkerFunc` constructor, but I don't think it's that bad.

I'll get a PR ready, but let me know if you folks like this approach.

Thanks, Eddy

### PR DESCRIPTION
**Pull request by ereyes01**
fixes #257 

Currently, the `ast.Visitor` interface has a single method- `Visit`- which takes as an argument the `Node` of each element of the AST as the `Walk` function visits it. When all the children in a node's AST subtree have been visited, the `Visit` method is called with a nil `Node` interface value. If you're keeping score and you really want to track when an AST node has been fully processed, this design isn't good enough- the AST can itself contain elements that are nil interface values, which are indistinguishable from "exiting" a node.

This PR modifies the `Visitor` interface such that it has two methods- `Enter` and `Exit`- to remove the ambiguity caused by the possibility that the AST has nil interface values.

Many more details can be found in #257

### CODE REVIEW

**Comment by ereyes01** [comment]
Thanks for the feedback. I've restored the example and I've modified the first comment with a summary of the problem and the change. Let me know if anything else needs to be done, thanks!

**Comment by ereyes01** [comment]
Bump... any other issues with this PR? Thanks...

File: ast/walk.go [hunk 1]
```diff
@@ -2,35 +2,29 @@ package ast
 
 import "fmt"
 
-// Visitor Visit method is invoked for each node encountered by Walk.
+// Visitor Enter method is invoked for each node encountered by Walk.
 // If the result visitor w is not nil, Walk visits each of the children
-// of node with the visitor v, followed by a call of v.Visit(nil).
+// of node with the visitor v, followed by a call of the End method.
```
Comment (darkliquid) [inline ast/walk.go]: Don't you mean a call of the `Exit` method here?

Comment (ereyes01) [inline ast/walk.go]: doh... yup. I'll fix the comment, thanks for catching.

**Pull request merged**